freedb/protocols.md

```diff
@@ -1,8 +1,311 @@
+used to store `_rid` metadata (which will be explained later).
+
+The first row will be used to store the column name of each column. The first column will be the `_rid` metadata column
+(which also has the `_rid` column name) and will be followed by the first column in the data schema, then followed by the
+second one, and so on. The next rows (second row and rows below that) will be used to store the records.
+
+Suppose the client schema looks like this (written in `PyFreeDB` model):
+```
+class Person(models.Model):
+    name = models.StringField() # 1st column
+    age = models.IntegerField() # 2nd column
+```
+
+and we have data below:
+```
+[
+    Person(name="A", age=1), # 1st row
+    Person(name="B", age=2), # 2nd row
+    Person(name="C", age=3), # 3rd row
+    Person(name="D", age=4)  # 4th row
+]
+```
+
+Then the actual data that we store in Google Sheet would look like this:
+| _rid | name | age |
+| ---- | ---- | --- |
+| 1    | A    | 1   |
+| 2    | B    | 2   |
+| 3    | C    | 3   |
+| 4    | D    | 4   |
+
+#### About `_rid` Metadata
+
+`_rid` column will be used to tell the index of a row. This metadata will make our Google Sheet Formula simpler
+so that we can utilize [Google Visualization API][GVizAPI] for our needs.
+
+⚠️ Even though `_rid` can be used as the row identifier (because `_rid` is unique between each row), we recommend the
+client to implement their own row identifier mechanism such as [UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier)
+or [Snowflake](https://en.wikipedia.org/wiki/Snowflake_ID) because we will not guarantee that `_rid` will not be
+reused out-of-the-box (to ensure the `_rid` is not reused, you need to implement soft-delete mechanism by your own).
+
+Consider this scenario:
+1. Client inserts Row A to the store, which will have `_rid=2`.
+2. Client deletes Row A from the store.
+3. Client Insert Row B to the store, which will reuse the same `_rid` as A i.e. B's `_rid` will be equal to `2`.
+
+We did not expose this field by default because if the client didn't aware of this behavior it can cause various bugs
+because of the invalid/dangling IDs.
+
+### Operations
+
+#### Insert
+
+Inserts new rows into the database.  The operation method should expect these parameters:
+| Parameter Name | Required | Remarks                                                   |
+| -------------- | -------- | --------------------------------------------------------- |
+| `rows`         | Yes      | List of row objects that will be inserted into the store. |
+
+
+Call [spreadsheets.values.append][AppendAPI]
+API with the following parameters:
+
+| Parameter Name              | Value                      |
+| --------------------------- | -------------------------- |
+| `spreadsheetId`             | `<current_spreadsheet_id>` |
+| `range`                     | `<data_range>`             |
```

Comment:
I think there should be a formula to decide on this data range, right? Otherwise it might be confusing to figure out. I believe data range should be part of the protocol.

Reply:
We have recommendation of what the `<data_range>` should be in L85. 

Technically there's a lot of value that we can use that works the same way, but the recommended way suggest to select the entire data column.

Reply:
Done